cpp/src/genetic/program.cu

```diff
@@ -0,0 +1,535 @@
+    node_probs[i] /= sum;
+  }
+
+  // Compute cumulative sum
+  std::partial_sum(node_probs.begin(),node_probs.end(),node_probs.begin());
+
+  // CUML_LOG_DEBUG("Current bound is %f",bound);
+  
+  start = std::lower_bound(node_probs.begin(),node_probs.end(),bound) - node_probs.begin();
+  end = start;
+
+  // Iterate until all function arguments are satisfied in current subtree
+  int num_args = 1;
+  while(num_args > end - start){
+    node curr;
+    curr = pnodes[end];
+    if(curr.is_nonterminal())num_args += curr.arity();
+    ++end;
+  }
+  
+  return std::make_pair(start,end);
+}
+
+int get_depth(program &p_out){
+  
+  auto depth = 0;
+  std::stack<int> arity_stack;
+  for(auto i = 0; i < p_out.len ; ++i){
+    
+    node curr(p_out.nodes[i]);
+
+    // Update depth
+    int sz = arity_stack.size();
+    depth = std::max(depth,sz);
+    
+    // Update stack
+    if(curr.is_nonterminal()){
+      arity_stack.push(curr.arity());
+    }
+    else{
+      // Only triggered for depth 0 node
+      if(arity_stack.empty())break;
+
+      int e = arity_stack.top();
+      arity_stack.pop();
+      arity_stack.push(e-1);
+
+      while(arity_stack.top() == 0){
+        arity_stack.pop();
+        if(arity_stack.empty())break;
+
+        e = arity_stack.top();
+        arity_stack.pop();
+        arity_stack.push(e-1);
+      }
+    }
+  }
+
+  // Set and return depth 
+  p_out.depth = depth;
+  return depth;
+}
+
+void build_program(program &p_out, const param &params,std::mt19937 &gen){
+  
+  // Define data structures needed for tree
+  std::stack<int> arity_stack;
+  std::vector<node> nodelist;
+  nodelist.reserve(1 << (MAX_STACK_SIZE));
+
+  // Specify RNGs
+  std::uniform_int_distribution<> dist_func(0,params.function_set.size()-1);
+  std::uniform_int_distribution<int> dist_depth(1, MAX_STACK_SIZE-1); 
+  std::uniform_int_distribution<int> dist_t(0,params.num_features);
+  std::uniform_int_distribution<int> dist_choice(0,params.num_features+params.function_set.size()-1);
+  std::uniform_real_distribution<float> dist_const(params.const_range[0],params.const_range[1]);
+  std::uniform_int_distribution<int> dist_U(0,1);
+
+  // MAX_STACK_SIZE can only handle tree of size MAX_STACK_SIZE - max(function_arity=2) + 1
+
+  // Initialize nodes
+  int max_depth = dist_depth(gen);
+  node::type func = params.function_set[dist_func(gen)];
+  node curr_node(func);
+  nodelist.push_back(curr_node);
+  arity_stack.push(curr_node.arity());
+  
+  init_method_t method = params.init_method;
+  if(method == init_method_t::half_and_half){
+    // Choose either grow or full for this tree
+    int ch = dist_U(gen);
+    method = ch == 0 ? init_method_t::grow : init_method_t::full;
+  }
+
+  // Fill tree
+  while(!arity_stack.empty()){
+    
+    int depth = arity_stack.size();
+    p_out.depth = std::max(depth,p_out.depth);
+    
+    int node_choice = dist_U(gen);
+
+    if((node_choice == 0 || method == init_method_t::full) && depth < max_depth){
```

Comment:
if `method != init_method_t::full` and `depth<max_depth`, this condition has a 50/50 chance of being true. `gplearn` seems to have a different probability [here](https://github.com/trevorstephens/gplearn/blob/ad57cb18caafdb02cca861aea712f1bf3ed5016e/gplearn/_program.py#L197). Is there a particular reason?

Reply:
I made it 50/50 to ideally not skew the selection distribution (when deciding between functions and features) if `init_method_t::grow` is selected. I thought the original distribution could potentially lead to shallower trees when more input features are present.

Reply:
And is there any accuracy/performance basis for that? Not an expert here, but the docs [here](https://gplearn.readthedocs.io/en/stable/intro.html#initialization) provides a distinction in tree nature based on nature of `function_set` and `features` provided. I'm talking about this paragraph in specific:
```
For the 'grow' method, nodes are chosen at random from both functions and terminals, allowing for smaller trees than init_depth specifies. This tends to grow asymmetrical trees as terminals can be chosen before the max depth is reached. If your dataset has a lot of variables, this will likely result in much smaller programs than init_depth specifies. Similarly, if you have very few variables and have chosen a large function set, you will likely see programs approaching the maximum depth specified by init_depth.
```
What do you think @teju85 ?

Reply:
Well, If we want to be more precise, Table 1 [here](https://dl.acm.org/doi/10.1109/4235.873237) suggests different problem domains have specific grow selection probabilities (for terminals vs non-terminals). We can simply add another parameter in `cuml::genetic::param` to represent this Bernoulli distribution and leave it to the user (the default probability still has to be considered).

Reply:
I think that will be a good idea 👍

Reply:
thanks for adding the `terminalRatio` param vimarsh